### PR TITLE
[AppKit] Add some of the missing NSAccessibility bits

### PR DESCRIPTION
**Pull request by iainx**

### CODE REVIEW

File: src/AppKit/NSAccessibility.cs [hunk 1]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+
+		public static void PostNotification (NSObject element, NSString notification)
+		{
+			NSAccessibilityPostNotification (element.Handle, notification.Handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityRoleDescription (IntPtr role, IntPtr subrole);
+
+		public static string RoleDescription (NSString role, NSString subrole)
```
Comment (spouliot) [inline src/AppKit/NSAccessibility.cs]: a method name needs a verb, e.g. `GetRoleDescription`

File: src/AppKit/NSAccessibility.cs [hunk 2]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
 			return NSAccessibilityPointInView (parentView, point);
 		}
+
+		[Mac (10,7)]
+		[DllImport (Constants.AppKitLibrary)]
+		static extern void NSAccessibilityPostNotificationWithUserInfo (IntPtr element, IntPtr notification, IntPtr userInfo);
+
+		public static void PostNotification (NSObject element, NSString notification, NSDictionary userInfo)
+		{
+			NSAccessibilityPostNotificationWithUserInfo (element.Handle, notification.Handle, userInfo.Handle);
```
Comment (spouliot) [inline src/AppKit/NSAccessibility.cs]: Any `null` arguments will cause NRE (common issue in several API here).

If `null` is allowed by the native API then `IntPtr.Zero` must be given to p/invoke.
If not then a null check with a ANE should be thrown.

File: src/AppKit/NSAccessibility.cs [hunk 3]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityRoleDescriptionForUIElement (IntPtr element);
+
+		public static string RoleDescription (NSObject element)
+		{
+			IntPtr handle = NSAccessibilityRoleDescriptionForUIElement (element.Handle);
+
+			if (handle == IntPtr.Zero)
+				return null;
```
Comment (spouliot) [inline src/AppKit/NSAccessibility.cs]: that check is not needed, `NSString.FromHandle` already _handle_ it

**Comment by spouliot** [comment]
@iainx thanks for the PR :) my 2 comments (method name and null check) needs to be applied globally. 

Also can you confirm that you ran the introspection tests with those changes ? that will spot some potential defects (including incorrect Apple documentation). If you did just reply here, thanks!

Also @chamons might have more feedback too...

**Comment by xamarin-release-manager** [comment]
Build success

File: src/AppKit/NSAccessibility.cs [hunk 4]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+
+			return NSString.FromHandle (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityActionDescription (IntPtr action);
+
+		public static string ActionDescription (NSString action)
```
Comment (chamons) [inline src/AppKit/NSAccessibility.cs]: Another "needs a verb" in name

File: src/AppKit/NSAccessibility.cs [hunk 5]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+		{
+			IntPtr handle = NSAccessibilityActionDescription (action.Handle);
+			return NSString.FromHandle (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredAncestor (IntPtr element);
+
+		public static NSObject UnignoredAncestor (NSObject element)
```
Comment (chamons) [inline src/AppKit/NSAccessibility.cs]: Again.

File: src/AppKit/NSAccessibility.cs [hunk 6]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+		{
+			var handle = NSAccessibilityUnignoredAncestor (element.Handle);
+			return new NSObject (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredDescendant (IntPtr element);
+
+		public static NSObject UnignoredDescendant (NSObject element)
```
Comment (chamons) [inline src/AppKit/NSAccessibility.cs]: Again.

File: src/AppKit/NSAccessibility.cs [hunk 7]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+		{
+			var handle = NSAccessibilityUnignoredDescendant (element.Handle);
+
+			return new NSObject (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredChildren (IntPtr originalChildren);
+
+		public static NSObject[] UnignoredChildren (NSArray originalChildren)
```
Comment (chamons) [inline src/AppKit/NSAccessibility.cs]: Again.

File: src/AppKit/NSAccessibility.cs [hunk 8]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+		{
+			var handle = NSAccessibilityUnignoredChildren (originalChildren.Handle);
+
+			return NSArray.ArrayFromHandle<NSObject> (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredChildrenForOnlyChild (IntPtr originalChild);
+
+		public static NSObject[] UnignoredChildrenForOnlyChild (NSObject originalChild)
```
Comment (chamons) [inline src/AppKit/NSAccessibility.cs]: Again.

**Comment by chamons** [comment]
I picked out the places where @spouliot's name change needs hit. Once you land those then LGTM.

File: src/appkit.cs [hunk 9]
```diff
@@ -21879,4 +22792,39 @@ public interface NSAccessibilityLayoutItem : NSAccessibilityGroup {
 		[Export ("setAccessibilityFrame:")]
 		void SetAccessibilityFrame (CGRect frame);
 	}
+
+	[Category, BaseType (typeof (NSObject))]
+	public interface NSObjectAccessibilityExtensions {
+		[Export ("accessibilityHitTest:")]
+		NSObject AccessibilityHitTest (CGPoint point);
```
Comment (spouliot) [inline src/appkit.cs]: same for binding categories, methods needs verbs :)

File: src/AppKit/NSAccessibility.cs [hunk 10]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+			return NSString.FromHandle (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredAncestor (IntPtr element);
+
+		public static NSObject UnignoredAncestor (NSObject element)
+		{
+			var handle = NSAccessibilityUnignoredAncestor (element.Handle);
+			return new NSObject (handle);
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs]: Use `Runtime.GetNSObject (handle)` instead, which does not create a new managed wrapper if one already exists.

File: src/AppKit/NSAccessibility.cs [hunk 11]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityUnignoredDescendant (IntPtr element);
+
+		public static NSObject UnignoredDescendant (NSObject element)
+		{
+			var handle = NSAccessibilityUnignoredDescendant (element.Handle);
+
+			return new NSObject (handle);
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs]: Use `Runtime.GetNSObject (handle)` instead.

File: src/AppKit/NSAccessibility.cs [hunk 12]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+			NSAccessibilityPostNotificationWithUserInfo (element.Handle, notification.Handle, userInfo.Handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern void NSAccessibilityPostNotification (IntPtr element, IntPtr notification);
+
+		public static void PostNotification (NSObject element, NSString notification)
+		{
+			NSAccessibilityPostNotification (element.Handle, notification.Handle);
+		}
```
Comment (mandel-macaque) [inline src/AppKit/NSAccessibility.cs]: I believe you need to check against null ptr here too. element and notification can be null and you will get a NRE.

File: src/AppKit/NSAccessibility.cs [hunk 13]
```diff
@@ -80,6 +89,93 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
+			return NSString.FromHandle (handle);
+		}
+
+		[DllImport (Constants.AppKitLibrary)]
+		static extern IntPtr NSAccessibilityRoleDescriptionForUIElement (IntPtr element);
+
+		public static string RoleDescription (NSObject element)
+		{
+			IntPtr handle = NSAccessibilityRoleDescriptionForUIElement (element.Handle);
+
```
Comment (mandel-macaque) [inline src/AppKit/NSAccessibility.cs]: A check is indeed needed for element AFAIK, else null.Handler will throw an exception.

File: src/appkit.cs [hunk 14]
```diff
@@ -21624,6 +21624,919 @@ partial interface NSAccessibilityAttributes {
...
+		NSString MarkerTypeDescriptionAttribute { get; }
+
+		[Mac (10, 7)]
+		[Field ("NSAccessibilityIdentifierAttribute")]
+		NSString IdentifierAttribute { get; }
+	}
+
+	[Static]
+	public interface NSAccessibilityFontKeys {
+		[Field ("NSAccessibilityFontNameKey")]
```
Comment (mandel-macaque) [inline src/appkit.cs]: What is the usage of this keys? Are they used to access data in an NSDictionary? Is it is the case we should use a strong dictionary.

File: src/appkit.cs [hunk 15]
```diff
@@ -21624,6 +21624,919 @@ partial interface NSAccessibilityAttributes {
...
+	[Static]
+	public interface NSWorkspaceAccessibilityNotifications {
+		[Mac (10, 10)]
+		[Field ("NSWorkspaceAccessibilityDisplayOptionsDidChangeNotification")]
+		NSString DisplayOptionsDidChangeNotification { get; }
+	}
+
+	[Static]
+	public interface NSAccessibilityNotificationUserInfoKeys {
+		[Field ("NSAccessibilityUIElementsKey")]
```
Comment (mandel-macaque) [inline src/appkit.cs]: Same question about the usage of the keys.

File: src/appkit.cs (hunk 14 above)
Comment (iainx) [inline src/appkit.cs]: Yes, the FontKeys are used with the NSDictionary returned using the FontTextAttribute.

File: src/appkit.cs (hunk 15 above)
Comment (iainx) [inline src/appkit.cs]: Again, yes, these are keys used in the NSDictionary user info passed to the PostNotification method

Comment (iainx) [inline src/appkit.cs]: But I think other custom keys can be used in the UserInfo dictionary too, so making a strong dictionary wouldn't be possible.

**Comment by iainx** [comment]
I've fixed all the bits mentioned and I've run the introspection tests.

I changed the NSObjectAccessibilityExtensions and NSWorkspaceAccessibilityExtensions to be a protocol instead of a category because it's really an informal protocol implemented as a category.

I also added all the obsolete methods because in standard Apple style they've obsoleted things without giving a better way of doing that task, so it looks like the obsolete methods will still be needed for certain things.

File: src/appkit.cs (hunk 14 above)
Comment (iainx) [inline src/appkit.cs]: But the NSDictionary is returned by the (obsoleted) GetAccessibilityAttributeValue method, which returns a generic NSObject, so we can't using the strong dictionary here either I'm afraid.

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by rolfbjarne** [comment]
My comments have been solved to my satisfaction, and I have nothing else 👍

File: src/appkit.cs [hunk 16]
```diff
@@ -21879,4 +22792,79 @@ public interface NSAccessibilityLayoutItem : NSAccessibilityGroup {
 		[Export ("setAccessibilityFrame:")]
 		void SetAccessibilityFrame (CGRect frame);
 	}
+
+	[Protocol, BaseType (typeof (NSObject))]
+	public interface NSObjectAccessibilityExtensions {
+		[Obsolete]
```
Comment (spouliot) [inline src/appkit.cs]: That's not really helpful for developers as it does not say why ? and what's the recommended replacement ?

In most cases that's better handled with something like:

`[Availability (Obsoleted = Platform.Mac_10_10, Message = "Use XXX instead")]`

as it tells when it was obsoleted and how to update the code (when possible). IIRC you mentioned some things were obsolete without an alternative and, if that's the case, then it should become the text, e.g. "No alternative is documented".

Also in some cases it's `Deprecated` and not `Obsoleted` that should be used (or both).

**Comment by iainx** [comment]
I've fixed the Obsolete attributes, directing people at the newer protocols, which is all Apple's docs do too, as there's no direct equivalent (it seems they're wanting to do things in a different way from 10.10)

I'm not sure what the difference between Obsoleted and Deprecated is so maybe I should have used Deprecated.

**Comment by xamarin-release-manager** [comment]
Build success

**Comment by chamons** [comment]
I agree with @rolfbjarne that everything looks reasonable now. Landing.

**Comment by spouliot** [comment]
@chamons dontlink-mac fails since this was merged https://wrench.internalx.com/Wrench/ViewLane.aspx?lane_id=1381&host_id=243&revision_id=768118

**Comment by iainx** [comment]
@spouliot @chamons yes, sorry the accessibilityAttributeNames shouldn't have the :
I'll make a new PR to fix that tomorrow

**Comment by iainx** [comment]
Is it possible to run these tests locally?

**Comment by chamons** [comment]
@iainx see https://github.com/xamarin/xamarin-macios/pull/143

and make run-mac-dontlink in tests dir

**Pull request merged**